### Solve system of equations using Gaussian elimination method:
$$
q - r + 2s = -3
$$

$$
4q + 4r -2s = 1
$$

$$
-2q +2r -4s = 6
$$

In [49]:
#=
Put into augmented matrix:

[1 -1  2 -3]
|4  4 -2  1|
[-2 2 -4  6]
=#

In [50]:
m = (Float64)[1 -1 2 -3;4 4 -2 1;-2 2 -4 6]

3×4 Matrix{Float64}:
  1.0  -1.0   2.0  -3.0
  4.0   4.0  -2.0   1.0
 -2.0   2.0  -4.0   6.0

In [51]:
# divide row 3 by 2
m[3,:] = m[3,:] ./ 2
display(m)

3×4 Matrix{Float64}:
  1.0  -1.0   2.0  -3.0
  4.0   4.0  -2.0   1.0
 -1.0   1.0  -2.0   3.0

In [52]:
# delete bottom row by adding 1st row to it
m[3,:] = m[3,:] .+ m[1,:]
display(m)

3×4 Matrix{Float64}:
 1.0  -1.0   2.0  -3.0
 4.0   4.0  -2.0   1.0
 0.0   0.0   0.0   0.0

In [53]:
# clear out 4 from 1st column
# row 2 = row 2 - 4 * row 1
m[2,:] = m[2,:] .- (4 .* m[1,:])
display(m)

3×4 Matrix{Float64}:
 1.0  -1.0    2.0  -3.0
 0.0   8.0  -10.0  13.0
 0.0   0.0    0.0   0.0

In [54]:
# divide row 2 by 8 to get pivot
m[2,:] = m[2,:] ./ 8
display(m)

3×4 Matrix{Float64}:
 1.0  -1.0   2.0   -3.0
 0.0   1.0  -1.25   1.625
 0.0   0.0   0.0    0.0

In [55]:
# add row 2 to row 1
m[1,:] .+= m[2,:]

4-element view(::Matrix{Float64}, 1, :) with eltype Float64:
  1.0
  0.0
  0.75
 -1.375

In [56]:
display(m)

3×4 Matrix{Float64}:
 1.0  0.0   0.75  -1.375
 0.0  1.0  -1.25   1.625
 0.0  0.0   0.0    0.0

#### Thus we can say
$$q + 0.75s = -1.375$$
$$r - 1.25s = 1.625$$

#### s is a free variable, which could have any value, and would not affect the rest of the system.

#### if
$$s = 0$$

#### then
$$ q = -1.375; r = 1.625$$

and if s is another value, q and r are affected